ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -149,21 +150,49 @@ typecheckModule (IdeDefer defer) hsc keep_lbls pm = do
 -- | Add a Hook to the DynFlags which captures and returns the
 -- typechecked splices before they are run. This information
 -- is used for hover.
-captureSplices :: DynFlags -> (DynFlags -> IO a) -> IO (a, [LHsExpr GhcTc])
+captureSplices :: DynFlags -> (DynFlags -> IO a) -> IO (a, SpliceInfo)
 captureSplices dflags k = do
-  splice_ref <- newIORef []
+  splice_ref <- newIORef mempty
   res <- k (dflags { hooks = addSpliceHook splice_ref (hooks dflags)})
   splices <- readIORef splice_ref
   return (res, splices)
   where
-    addSpliceHook :: IORef [LHsExpr GhcTc] -> Hooks -> Hooks
+    addSpliceHook :: IORef SpliceInfo -> Hooks -> Hooks
     addSpliceHook var h = h { runMetaHook = Just (splice_hook var) }
 
-    splice_hook :: IORef [LHsExpr GhcTc] -> MetaRequest -> LHsExpr GhcTc -> TcM MetaResult
-    splice_hook var mr e = do
-      liftIO $ modifyIORef var (e:)
-      pprTraceM "expr" (ppr e)
-      defaultRunMeta mr e
+    splice_hook :: IORef SpliceInfo -> MetaHook TcM
+    splice_hook var metaReq e = case metaReq of
+        (MetaE f) -> do
+            expr' <- metaRequestE defaultRunMeta e
```

Comment:
You should use the original hook instead of `defaultRunMeta` so that hooks set elsewhere aren't silently overwritten.

Reply:
Done. Thanks for pointing out!